collections-contrib/src/main/scala/strawman/collection/decorators/MapDecorator.scala

```diff
@@ -20,4 +29,108 @@ trait MapDecorator[K, V] {
     b.result()
   }
 
+  /**
+    * Combines the entries of `this` Map with the entries of `that` Map that have the same key,
+    * tupling their values. `xs.zipByKey(ys)` is a shorthand for `xs.zipByKeyWith(ys)((_, _))`.
+    *
+    * @param that Other Map
+    * @param bf Builder driven by the type of `this` Map
+    * @tparam W Type of values of the other Map (e.g. `Int`, `String`)
+    * @tparam C Type of the result (e.g. `Map[Int, (String, Boolean)]`, `TreeMap[String, (Int, Int)]`)
+    * @return A Map that associates all the keys `k` contained by both `this` and `that` to pairs `(v, w)` where `v`
+    *         is the value associated by `this` to `k` and `w` the value associated by `that` to `k`
+    */
+  def zipByKey[W, C](that: Map[K, W])(implicit bf: BuildFrom[`this`.type, (K, (V, W)), C]): C =
+    zipByKeyWith(that)((_, _))
+
+  /** Alias for `zipByKey` */
+  @`inline` final def join[W, C](that: Map[K, W])(implicit bf: BuildFrom[`this`.type, (K, (V, W)), C]): C = zipByKey(that)
+
+  /**
+    * @return A Map associating all the keys from `this` and `that` with values returned by the partial function
+    *         `f`, when this one is defined.
+    *
+    * @param that Map to merge
+    * @param f Combination function
+    * @param bf Builder driven by the type of `this` Map
+    * @tparam W Type of values of the other Map (e.g. `Int`, `String`)
+    * @tparam X Type of values of the resulting Map
+    * @tparam C Type of the result (e.g. `Map[Int, (String, Option[Boolean])]`)
+    */
+  def mergeByKeyWith[W, X, C](that: Map[K, W])(f: PartialFunction[(Option[V], Option[W]), X])(implicit bf: BuildFrom[`this`.type, (K, X), C]): C = {
```

Comment:
I considered introducing a type `EitherOrBoth[A, B]` like so:

~~~ scala
type EitherOrBoth[+A, +B] = Either[(A, B), Either[A, B]]
~~~

But at use site it was less convenient than a pair of options.

Reply:
> it was less convenient

@julienrf How so?

Reply:
First, the goal of `EitherOrBoth` is to not represent meaningless states like `(None, None)`, which could be useful to avoid non exhaustiveness warnings. But in practice we always take as parameter a `PartialFunction`, so exhaustiveness doesn’t matter.

Also, with `(Option[V], Option[W])` we can deal with the cases where `V` is defined in one case: `case (Some(v), _) => …`. Whereas with `EitherOrBoth[V, W]` you would have to write two cases: `case Left((v, _)) => … case Right(Left(v)) => …`.

Reply:
Letting users write `{case (None,None) => ...}` as if it made sense is not great either.

`EitherOrBoth` should have an extractor that looks for a `Left` value whether there is a `Right` one or not: something like `AtLeastLeft(x)`, though the `Left` extractor itself could also have this semantics (would it be too surprising?).

It looks like you have tried encoding `EitherOrBoth[A,B]` as `Either[(A,B),Either[A,B]]`. It's not surprising to me that this was annoying to use, compared to a proper `EitherOrBoth` data type 😉

Reply:
Yeah, good point :)